server/src/unifyfs_tree.c

```diff
@@ -989,3 +989,268 @@ static void unlink_response_rpc(hg_handle_t handle)
+
+/******************************************************************
+ * metaset 
+ ******************************************************************/
+
+static int rpc_invoke_metaset_request(int rank, unifyfs_coll_state_t* st)
+{
+    int rc = (int)UNIFYFS_SUCCESS;
+
+    /* get address for specified server rank */
+    hg_addr_t addr = glb_servers[rank].margo_svr_addr;
+
+    /* get handle to rpc function */
+    hg_handle_t handle;
+    hg_return_t hret = margo_create(unifyfsd_rpc_context->svr_mid, addr,
+        unifyfsd_rpc_context->rpcs.metaset_request_id, &handle);
+    assert(hret == HG_SUCCESS);
+
+    /* fill in input struct */
+    unifyfs_file_attr_t *attr = &st->attr;
+    metaset_request_in_t in;
+    in.root   = (int32_t)st->root;
+    in.tag    = (int32_t)st->tag;
+    in.create = (int32_t)st->create;
+    in.attr   = *attr;
+
+    /* call rpc function */
+    hret = margo_forward(handle, &in);
+    assert(hret == HG_SUCCESS);
+
+    /* wait for rpc output */
+    metaset_request_out_t out;
+    hret = margo_get_output(handle, &out);
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+static int rpc_invoke_metaset_response(int rank, unifyfs_coll_state_t* st)
+{
+    int rc = (int)UNIFYFS_SUCCESS;
+
+    /* get address for specified server rank */
+    hg_addr_t addr = glb_servers[rank].margo_svr_addr;
+
+    /* get handle to rpc function */
+    hg_handle_t handle;
+    hg_return_t hret = margo_create(unifyfsd_rpc_context->svr_mid, addr,
+        unifyfsd_rpc_context->rpcs.metaset_response_id, &handle);
+    assert(hret == HG_SUCCESS);
+
+    /* fill in input struct */
+    metaset_response_in_t in;
+    in.tag      = (int32_t) st->parent_tag;
+    in.err      = (int32_t) st->err;
+
+    /* call rpc function */
+    hret = margo_forward(handle, &in);
+    assert(hret == HG_SUCCESS);
+
+    /* wait for rpc output */
+    metaset_response_out_t out;
+    hret = margo_get_output(handle, &out);
+    assert(hret == HG_SUCCESS);
+
+    /* decode response */
+    rc = (int) out.ret;
+
+    /* free resources */
+    margo_free_output(handle, &out);
+    margo_destroy(handle);
+
+    return rc;
+}
+
+static void metaset_response_forward(unifyfs_coll_state_t* st)
+{
+    printf("%d: BUCKEYES response_forward (metaset)\n", glb_pmi_rank);
+    fflush(stdout);
+    /* get tree we're using for this operation */
+    unifyfs_tree_t* t = &st->tree;
+
+    /* get info for tree */
+    int parent       = t->parent_rank;
+    int child_count  = t->child_count;
+    int* child_ranks = t->child_ranks;
+
+    /* send up to parent if we have gotten all replies */
+    if (st->num_responses == child_count) {
+        /* metaset the file */
+        int ret = gfid2ext_tree_metaset(&glb_gfid2ext, st->gfid, st->create,
+                                        &st->attr);
+
```

Comment:
We should update the st->err field if we failed to insert/update the metadata here.